bless/bless.1
.TH BLESS 1 "June 4, 2001" "Apple Computer, Inc."
.SH NAME
bless \- set boot options for volumes and Open Firmware
.SH SYNOPSIS
.B bless
-info
.I mountpoint
[
-plist
]
[
-verbose
] [
-bootBlocks
]

.B bless
[
-bootinfo
.I "arg"
] [
-bootBlocks
] [
-noexec
] [
-folder
.I folder
] [
-folder9
.I folder
] [
-quiet
] [
-setOF
] [
-use9
] [
-verbose
]
.SH DESCRIPTION
The
.I bless
command modifies on-disk an Open Firmware-based parameters used during the
boot sequence. It allows folders the be \(lqblessed\(rq in the volume header, which
allows OF to find the loader program (usually BootX).
.I bless
also sets Open Firmware variables to boot off the volume specified.
.PP
In the first usage, the following flags apply:
.TP
.B \-bootBlocks
Display important data from the boot blocks for the specified volume
.TP
.B \-info mountpoint
Choose the volume mounted at
.I mountpoint
to print information about.
.TP
.B \-plist
Display the requested information as a Property List using the
CoreFoundation API.
.TP
.B \-verbose
Be verbose while collecting data and accessing the volume.
.PP
Root privelege is not required to collect this information, although you must
have read access to the volume.
.PP
The second form of the command can take:
.TP
.B \-bootinfo arg
Create a new BootX file in the folder specified by
.B \-folder.
.B arg
should point to a
.I bootx.bootinfo
file, most commonly /System/Library/CoreServices/bootx.bootinfo
.TP
.B \-bootBlocks
Create boot blocks for the specified volume by extracting them from the
'boot' 1 resource of the
.I System
file contained in
.B \-folder9. Write these boot blocks to the volume.
.TP
.B \-noexec
Simulate execution of
.B bless
without performing any permanent changes to the system.
This flag is used by default when
.B bless
is run by a non-root user.
.TP
.B \-folder folder
The Mac OS X/Darwin folder to by blessed. This folder should have
a valid BootX file (unless
.B \-bootinfo
was also specified to create a new one). The folder's volume is what
all disk operations will be performed on. This is most commonly
/System/Library/CoreServices
.TP
.B \-folder9 folder
The Mac OS 9 System folder to bless. This folder should have a valid
System file and Finder. If
.I CarbonCore
is available, the
.B \-bootBlocks
option is also available. 
.TP
.B \-quiet
Don't print out any information except errors
.TP
.B \-setOF
Using the volume where
.B \-folder
resides (or both
.B \-folder
and
.B \-folder9,
if both were specified), set Open Firmware variables to boot off this device.
.TP
.B \-use9
If both
.B \-folder
and
.B \-folder9
were specified, prefer
.B \-folder9
as the primary blessed folder.
.TP
.B \-verbose
Be verbose while collecting data and accessing the volume.

.SH EXAMPLES

Darwin and Mac OS 9 are on seperate partitions, and want to boot into OS 9:

bless -folder9 "/Volumes/Mac OS 9/System Folder" -setOF

Darwin and Mac OS 9 on the same partition, and want to boot into OS 9:

bless -folder "/System/Library/CoreServices" -folder9 "/Volumes/Mac OS 9/System Folder"

Darwin and Mac OS X  on seperate partitions, and want to boot into Mac OS X:

bless -folder "/Volumes/Mac OS X/System/Library/CoreServices" -setOF

BootX file needs to be created (for creating a bootable partition):

bless -folder "/Volumes/Test/System/Library/CoreServices" -bootinfo /usr/standalone/ppc/bootx.bootinfo

.SH BUGS
.B bless
is only intelligent enough to deal with HFS+ volumes for most cases. It will
detect external booters, but other functionality may not work and has not been well-tested.





X11/ChangeLog
2003-09-12  John Harper  <jsh@apple.com>

	* xc/programs/Xserver/hw/apple/X11Application.m: revert fix for
	3363492 and add another check for RootlessKnowsWindowNumber ()
	to fix the original bug (3420745)

2003-09-10  John Harper  <jsh@apple.com>

	* xc/programs/Xserver/hw/apple/bundle-main.c (sigchld_handler):
	loop until no more zombies exist (3414442)

2003-08-08  John Harper  <jsh@apple.com>

	* xc/programs/Xserver/hw/apple/quartz.c,
	xc/programs/Xserver/hw/apple/X11Application.m,
	xc/programs/Xserver/hw/apple/X11Application.h: move calling
	HideMenuBar and ShowMenuBar onto the main UI thread (3359753)

	* xc/lib/GL/apple/dri_driver.h, xc/lib/GL/apple/dri_driver.c:
	few more MT changes, be even more careful about when contexts
	are updated

2003-08-07  John Harper  <jsh@apple.com>

	* xc/lib/GL/apple/dri_driver.h, xc/lib/GL/apple/dri_driver.c:
	Large set of changes to make this code handle multiple threads
	sanely. Changes include adding a mutex for each screen, and
	only ever touching CGL contexts when we know they're bound to
	the current thread or no thread. This involves patching gli
	dispatch table to cause glViewport to call back into DRI code
	so that contexts can be updated on their own thread. May be
	occasions where glXSwapBuffers fails if no suitable context can
	be found to issue the swap on (3340377) 

	* xc/programs/Xserver/hw/apple/quartz.c: quartzServerVisible
	variable needs to start out as false (3366845)

2003-08-04  John Harper  <jsh@apple.com>

	* xc/programs/Xserver/hw/apple/X11Application.m: move the call
	to RootlessKnowsWindowNumber () into the event handler so that
	we don't activate X on the broken menu-mouse-down event
	(3363492)

	* RELEASE-NOTES: updated to reflect that X11 is now part of the
	OS and no longer a preview release (3361286)

2003-07-29  John Harper  <jsh@apple.com>

	* xc/lib/GL/glx/glxext.c (glXMakeCurrent): unlock display after
	sending GLXMakeCurrent request when changing from indirect to
	direct context (3351489, sent upstream via Torrey)

2003-07-25  John Harper  <jsh@apple.com>

	* xc/programs/Xserver/hw/apple/X11Application.m (sendEvent):
	when in fullscreen mode, don't pass ApplicationDidReactivate
	events through to AppKit (3341360)

2003-07-23  John Harper  <jsh@apple.com>

	* xc/lib/GL/apple/dri_dispatch.c (XAppleDRIGetIndirectContext):
	make the context we allocate 1k bigger than we think it needs
	to be, in case GL changes

2003-07-22  John Harper  <jsh@apple.com>

	* xc/lib/GL/apple/dri_dispatch.c (XAppleDRIGetIndirectContext):
	allocate the CGLContext used for dispatch indirect rendering
	ourselves, avoids ever needing to contact the window server
	(3126715)

2003-07-18  John Harper  <jsh@apple.com>

	* xc/programs/Xserver/hw/apple/dri-surface.c: fixed use of
	pixmap private index

	* xc/programs/Xserver/hw/apple/quartz.c (QuartzSetFullscreen):
	use xp_disable_hot_keys instead of CG spi (3221162)

	* xc/programs/Xserver/hw/apple/rootless-screen.c,
	xc/programs/Xserver/hw/apple/rootless-common.h,
	xc/programs/Xserver/hw/apple/rootless-common.c: use
	RootlessFinishedDrawing () to call new function
	RootlessMayRedisplay (). After a certain amount of time since
	the last update it will cause a flush (3270678)

	* xc/programs/Xserver/hw/apple/rootless-screen.c: clean-up

2003-07-17  John Harper  <jsh@apple.com>

	* xc/programs/Xserver/hw/apple/rootless-window.c
	(RootlessPaintWindowBackground): prevent trying to draw in
	surfaces here

	* xc/programs/Xserver/hw/apple/rootless-gc.c: call
	RootlessFinishedDrawing at the end of each rendering primitive

	* xc/programs/Xserver/hw/apple/rootless-common.h,
	xc/programs/Xserver/hw/apple/rootless-common.c: use the
	DRIStartDrawing functions etc to test if the drawable is a
	surface.
	(RootlessFinishedDrawing): new function, should be called after
	drawing has finished

	* xc/programs/Xserver/hw/apple/dristruct.h,
	xc/programs/Xserver/hw/apple/dri-surface.h,
	xc/programs/Xserver/hw/apple/dri-surface.c (DRIStartDrawing,
	DRIStopDrawing, DRIDamageRegion, DRISynchronizeDrawable,
	DRISynchronize): functions to support Xlib rendering on
	accelerated surfaces. Only supported on 10.3, and they won't do
	anything by default

	* xc/programs/Xserver/hw/apple/dri-surface.c
	(DRICreateSurface): support Pixmap drawables (3122756, 10.3
	only)

	* xc/programs/Xserver/hw/apple/dri-surface.c: removed check for
	cg from SUJagGreen

	* xc/programs/Xserver/GL/apple/indirect.c: support offscreen
	drawables (by removing the code that stopped them being
	supported, dri-surface.c does the real work) (3122756)

	* xc/programs/Xserver/hw/apple/quartz.h,
	xc/programs/Xserver/hw/apple/quartz.c,
	xc/programs/Xserver/hw/apple/quartz-server.c,
	xc/programs/Xserver/hw/apple/X11Application.m,
	xc/programs/Xserver/hw/apple/X11Application.h: added xp_options
	default, lets you pass extra init options to Xplugin

	* xc/programs/Xserver/hw/apple/*: fixed gcc 3.3 warnings

2003-07-16  John Harper  <jsh@apple.com>

	* xc/programs/Xserver/hw/apple/x-list.c,
	xc/programs/Xserver/hw/apple/x-list.h (x_list_pop): new
	function

	* xc/programs/Xserver/hw/apple/x-hash.c: fix gcc 3.3 warning

2003-07-07  John Harper  <jsh@apple.com>

	* xc/programs/Xserver/hw/apple/X11Application.m: don't pass
	flags changed events to X11 when Appkit is active (3316333)

2003-07-02  John Harper  <jsh@apple.com>

	* xc/programs/Xserver/hw/apple/quartz.c (QuartzRunClient):
	don't quote commands, it stops multi-word commands working

	* xc/programs/Xserver/hw/apple/X11Controller.m: updated quit
	alert text (3313133)

	* xc/programs/Xserver/hw/apple/Xquartz.man: added more defaults

	* xc/programs/Xserver/hw/apple/bundle-main.c (main): detach

bless/bless.c
/*
 * Copyright (c) 2001-2020 Apple Inc. All Rights Reserved.
 *
 * @APPLE_LICENSE_HEADER_START@
 * 
 * This file contains Original Code and/or Modifications of Original Code
 * as defined in and that are subject to the Apple Public Source License
 * Version 2.0 (the 'License'). You may not use this file except in
 * compliance with the License. Please obtain a copy of the License at
 * http://www.opensource.apple.com/apsl/ and read it before using this
 * file.
 * 
 * The Original Code and all software distributed under the License are
 * distributed on an 'AS IS' basis, WITHOUT WARRANTY OF ANY KIND, EITHER
 * EXPRESS OR IMPLIED, AND APPLE HEREBY DISCLAIMS ALL SUCH WARRANTIES,
 * INCLUDING WITHOUT LIMITATION, ANY WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE, QUIET ENJOYMENT OR NON-INFRINGEMENT.
 * Please see the License for the specific language governing rights and
 * limitations under the License.
 * 
 * @APPLE_LICENSE_HEADER_END@
 */
/*
 *  bless.c
 *  bless
 *
 *  Created by Shantonu Sen <ssen@apple.com> on Wed Nov 14 2001.
 *  Copyright (c) 2001-2007 Apple Inc. All Rights Reserved.
 *
 *  $Id: bless.c,v 1.85 2006/07/17 22:19:05 ssen Exp $
 *
 */

#include <stdlib.h>
#include <stdio.h>
#include <string.h>
#include <stdarg.h>
#include <unistd.h>
#include <getopt.h>
#include <paths.h>
#include <err.h>
#include <sys/stat.h>
#include <sys/mount.h>
#include <sys/types.h>
#include <sys/sysctl.h>

#include "enums.h"
#include "structs.h"

#include "bless.h"
#include "bless_private.h"
#include "protos.h"

struct clarg actargs[klast];

/*
 * To add an option, allocate an enum in enums.h, add a getopt_long entry here,
 * add to main(), add to usage and man page
 */

/* options descriptor */
static struct option longopts[] = {
{ "apfsdriver",     required_argument,      0,              kapfsdriver},
{ "alternateos",    required_argument,      0,              kalternateos},
{ "alternateOS",    required_argument,      0,              kalternateos},
{ "allowUI",        no_argument,            0,              kallowui},
{ "bootinfo",       optional_argument,      0,              kbootinfo},
{ "bootefi",        optional_argument,      0,              kbootefi},
{ "booter",         required_argument,      0,              kbooter },
{ "create-snapshot",no_argument,            0,              kcreatesnapshot },
{ "device",         required_argument,      0,              kdevice },
{ "firmware",       required_argument,      0,              kfirmware },
{ "file",           required_argument,      0,              kfile },
{ "folder",         required_argument,      0,              kfolder },
{ "folder9",        required_argument,      0,              kfolder9 },
{ "getBoot",        no_argument,            0,              kgetboot },
{ "getboot",        no_argument,            0,              kgetboot },
{ "help",           no_argument,            0,              khelp },
{ "info",           optional_argument,      0,              kinfo },
{ "kernel",         required_argument,      0,              kkernel },
{ "kernelcache",    required_argument,      0,              kkernelcache },
{ "label",          required_argument,      0,              klabel },
{ "labelfile",      required_argument,      0,              klabelfile },
{ "last-sealed-snapshot",no_argument,       0,              klastsealedsnapshot },
{ "legacy",         no_argument,            0,              klegacy },
{ "legacydrivehint",required_argument,      0,              klegacydrivehint },
{ "mkext",          required_argument,      0,              kmkext },
{ "mount",          required_argument,      0,              kmount },
{ "netboot",        no_argument,            0,              knetboot},
{ "nextonly",       no_argument,            0,              knextonly},
{ "noapfsdriver",   no_argument,            0,              knoapfsdriver},
{ "openfolder",     required_argument,      0,              kopenfolder },
{ "options",        required_argument,      0,              koptions },
{ "passpromt",      no_argument,            0,              kpasspromt },
{ "payload",        required_argument,      0,              kpayload },
{ "personalize",    no_argument,            0,              kpersonalize },
{ "plist",          no_argument,            0,              kplist },
{ "quiet",          no_argument,            0,              kquiet },
{ "recovery",        no_argument,           0,              krecovery },
{ "reset",          no_argument,            0,              kreset },
{ "save9",          no_argument,            0,              ksave9 },
{ "saveX",          no_argument,            0,              ksaveX },
{ "server",         required_argument,      0,              kserver },
{ "setBoot",        no_argument,            0,              ksetboot },
{ "setboot",        no_argument,            0,              ksetboot },
{ "setOF",          no_argument,            0,              ksetboot }, // legacy option name
{ "shortform",      no_argument,            0,              kshortform },
{ "startupfile",    required_argument,      0,              kstartupfile },
{ "stdinpass",      no_argument,            0,              kstdinpass },
{ "unbless",        required_argument,      0,              kunbless },
{ "user",           required_argument,      0,              kuser },
{ "use9",           no_argument,            0,              kuse9 },
{ "use-tdm-as-external", no_argument,       0,              kusetdmasexternal },
{ "verbose",        no_argument,            0,              kverbose },
{ "version",        no_argument,            0,              kversion },
{ "snapshot",       required_argument,      0,              ksnapshot },
{ "snapshotname",   required_argument,      0,              ksnapshotname },
{ 0,            0,                      0,              0 }
};

extern char *optarg;
extern int optind;
extern char blessVersionString[];
char blessVersionNumString[64];
BLPreBootEnvType firmwareType;

BLPreBootEnvType getPrebootType(void) {
    return firmwareType;
}

static void blessHandleArgument(char * argv[], struct clarg *actarg, struct option *opt)
{
    if(actarg->present) {
        warnx("Option \"%s\" already specified", opt->name);
        usage_short();
        return;
    } else {
        actarg->present = 1;
    }

    switch(opt->has_arg) {
        case no_argument:
            actarg->hasArg = 0;
            break;
        case required_argument:
            actarg->hasArg = 1;
            strlcpy(actarg->argument, optarg, sizeof(actarg->argument));
            break;
        case optional_argument:
            if(argv[optind] && argv[optind][0] != '-') {
                actarg->hasArg = 1;
                strlcpy(actarg->argument, argv[optind], sizeof(actarg->argument));
            } else {
                actarg->hasArg = 0;
            }
            break;
    }
}


int main (int argc, char * argv[])
{

    int ch, longindex;
    BLContext context;
    struct blesscon bcon;

	strlcpy(blessVersionNumString, strrchr(blessVersionString, '-') + 1, sizeof blessVersionNumString);
	*strchr(blessVersionNumString, '\n') = '\0';
	
    bcon.quiet = 0;
    bcon.verbose = 0;

    context.version = 0;
    context.logstring = blesslog;
    context.logrefcon = &bcon;
    
    if (BLGetPreBootEnvironmentType(&context, &firmwareType)) {
        errx(1, "Could not determine firmware environment");
    }

    if(argc == 1) {
        usage_short();
    }
    
    if(getenv("BL_PRINT_ARGUMENTS")) {
        int i;
        for(i=0; i < argc; i++) {
            fprintf(stderr, "argv[%d] = '%s'\n", i, argv[i]);
        }
    }
    

    
    while ((ch = getopt_long_only(argc, argv, "", longopts, &longindex)) != -1) {
        if (firmwareType == kBLPreBootEnvType_iBoot) {
            switch(ch) {
                case khelp:
                    usage();
                    break;
                case kquiet:
                    break;
                case kverbose:
                    bcon.verbose = 1;
                    break;
                case kversion:
                    printf("%s\n", blessVersionNumString);
                    exit(0);
                    break;
                case kapfsdriver:
                    errx(1, "The 'apfsdriver' option is not supported on Apple Silicon devices.");
                    break;
                case kalternateos:
                    warnx("The 'alternateos/alternateOS' option is deprecated\n");
                    break;
                case kbootinfo:
                    errx(1, "The 'bootinfo' option is not supported on Apple Silicon devices.");
                    break;
                case kbooter:
                    errx(1, "The 'booter' option is not supported on Apple Silicon devices.");
                    break;
                case kfirmware:
                    errx(1, "The 'firmware' option is not supported on Apple Silicon devices.");
                    break;
                case kfolder9:
                    errx(1, "The 'folder9' option is not supported on Apple Silicon devices.");
                    break;
                case kkernel:
                    errx(1, "The 'kernel' option is not supported on Apple Silicon devices.");
                    break;
                case kkernelcache:
                    errx(1, "The 'kernelcache' option is not supported on Apple Silicon devices.");
                    break;
                case klegacy:
                case klegacydrivehint:
                    errx(1, "The 'legacy' and 'legacydriverhint' options are not supported on Apple Silicon devices.");
                    break;
                case kmkext:
                    errx(1, "The 'mkext' option is not supported on Apple Silicon devices.");
                    break;
                case knetboot:
                    errx(1, "The 'netboot' option is not supported on Apple Silicon devices.");
                    break;
                case kopenfolder:
                    errx(1, "The 'openfolder' is not supported on Apple Silicon devices.");
                    break;
                case kpayload:
                    errx(1, "The 'payload' option is not supported on Apple Silicon devices.");
                    break;
                case kserver:
                    errx(1, "The 'server' option is not supported on Apple Silicon devices.");
                    break;
                case ksave9:
                    warnx("The 'save9' option is deprecated\n");
                    break;
                case kuse9:
                    warnx("The 'use9' option is deprecated\n");
                    break;
                case '?':
                case ':':
                    usage_short();
                    break;
                default:
                    blessHandleArgument(argv, &actargs[ch], &longopts[longindex]);
                    break;
            }
        } else {
            switch(ch) {
                case khelp:
                    usage();
                    break;
                case kquiet:
                    break;
                case kverbose:
                    bcon.verbose = 1;
                    break;
                case kversion:
                    printf("%s\n", blessVersionNumString);
                    exit(0);
                    break;
                case kalternateos:
                    warnx("The 'alternateos/alternateOS' option is deprecated\n");
                    break;
                case kbootinfo:
                    warnx("The 'bootinfo' option is deprecated\n");
                    break;
                case kfolder9:
                    warnx("The 'folder9' option is deprecated\n");
                    break;
                case kopenfolder:
                    warnx("The 'openfolder' option is deprecated\n");
                    break;
                case kpayload:
                    actargs[ch].present = 1;
                    break;
                case ksave9:
                    warnx("The 'save9' option is deprecated\n");
                    break;
                case kuse9:
                    warnx("The 'use9' option is deprecated\n");
                    break;
                case '?':
                case ':':
                    usage_short();
                    break;
                default:
                    blessHandleArgument(argv, &actargs[ch], &longopts[longindex]);
                    break;
            }
        }
    }
    argc -= optind;
    argc += optind;
    
    /* There are 5 public modes of execution: info, device, folder, netboot, unbless
     * There is 1 private mode: firmware
     * These are all one-way function jumps.
     */
    if (firmwareType == kBLPreBootEnvType_iBoot) {
        /* External TDM devices are treated as standalone devices.
         * In order to use a TDM device as an external media to boot the current Apple silicon device use
         * the '--use-tdm-as-external' option.
         * For external/removable devices which are not TDM devices the following rules apply:
         * '--folder', '--file' and '--info' are used to boot/bless and get information based on x86 architecture
         * with the intent that this external/removable device will be used for booting an x86 Mac.
         * All other options are used to boot/bless and get information based on arm64e architecture
         * with the intent that this external/removable device will be used for booting this Apple Silicon Mac.
         */
        bool tdm = false;
        bool external = false;
        bool removable = false;
        char bsdName[BSD_NAME_SIZE];
        int ret = 0;

        if (!actargs[kgetboot].present) {
            if (actargs[kinfo].present) {
                ret = BLGetCommonMountPoint(&context, actargs[kinfo].argument, "", actargs[kmount].argument);
            } else if (actargs[kunbless].present) {
                ret = BLGetCommonMountPoint(&context, actargs[kunbless].argument, "", actargs[kmount].argument);
            } else if (actargs[kfolder].present) {
                ret = extractMountPoint(&context, actargs);
            }
            if (ret) {
                return ret;
            }
            if (actargs[kdevice].present) {
                strlcpy(bsdName, actargs[kdevice].argument + strlen(_PATH_DEV), sizeof(bsdName));
            } else if ((ret = extractDiskFromMountPoint(&context, actargs[kmount].argument, bsdName, BSD_NAME_SIZE)) != 0) {
                blesscontextprintf(&context, kBLLogLevelError, "Could not extract BSD name from mount point %s\n", actargs[kmount].argument);
                return ret;
            }

            ret = isMediaTDM(&context, bsdName, &tdm);
            if (ret) {
                return ret;
            }
            ret = isMediaExternal(&context, bsdName, &external);
            if (ret) {
                return ret;
            }
            ret = isMediaRemovable(&context, bsdName, &removable);
            if (ret) {
                return ret;
            }
        }

        /* '--setboot' and '--nextonly' relevant only for booting current internal Apple Silicon device
         * '--folder' is used to boot device based on Intel Atchitecture. Therefore those options
         * are not relevant.
         */
        if (actargs[kfolder].present) {
            if (actargs[ksetboot].present) {
                errx(1, "For Apple Silicon Macs, 'setboot' option is not supported in Folder mode");
            }
            if (actargs[knextonly].present) {
                errx(1, "For Apple Silicon Macs, 'nextonly' option is not supported in Folder mode");
            }
        }

        /* Handle Info mode (--info, --getBoot):
         * TDM is also considered an external devices
         */
        if (actargs[kgetboot].present ||
            (actargs[kinfo].present && external) ||
            (actargs[kinfo].present && removable)) {
            /* If it was requested, print out the Finder Info words */
            return modeInfo(&context, actargs);
        } else if (actargs[kinfo].present) {
            errx(1, "For Apple Silicon Macs, the 'info' option is only supported for external devices.");
        }

        /* Handle TDM devices (Device mode, Unbless mode, File/Folder mode, Mount mode,
         * Info mode is handled above):
         */
        if (tdm && !actargs[kusetdmasexternal].present) {
            if (actargs[kdevice].present) {
                return modeDevice(&context, actargs);
            }
            if (actargs[kunbless].present) {
                return modeUnbless(&context, actargs);
            }
            return modeFolder(&context, actargs);
        }

        /* Handle external and removable devices not in TDM (Folder mode): */
        if (!tdm && (external || removable)) {
            if (actargs[kfolder].present) {
                return modeFolder(&context, actargs);
            }
        }

        if (actargs[kunbless].present) {
            errx(1, "For Apple Silicon Macs, the 'unbless' option is only supported for Intel architecture based devices in TDM");
        }
        if (actargs[kfolder].present) {
            errx(1, "For Apple Silicon Macs, the 'folder' option is only supported for external devices");
        }
        if (actargs[kfile].present) {
            errx(1, "For Apple Silicon Macs, the 'file' option is supported for external devices in Folder mode only");
        }

        /* Handle AS devices (Device mode, Mount mode,
         * Info mode is handled above):
         */
        return blessViaBootability(&context, actargs);

    } else {
        /* If it was requested, print out the Finder Info words */
        if(actargs[kinfo].present || actargs[kgetboot].present) {
            return modeInfo(&context, actargs);
        }

        if(actargs[kdevice].present) {
            return modeDevice(&context, actargs);
        }

        if(actargs[kfirmware].present) {
            return modeFirmware(&context, actargs);
        }

        if(actargs[knetboot].present) {
            return modeNetboot(&context, actargs);
        }

        if (actargs[kunbless].present) {
            return modeUnbless(&context, actargs);
        }

        /* default */
        return modeFolder(&context, actargs);
    }
}



// note that libbless has its own (similar) contextprintf()
int blesscontextprintf(BLContextPtr context, int loglevel, char const *fmt, ...) {
    int ret;
    char *out;
    va_list ap;

    va_start(ap, fmt);
    ret = vasprintf(&out, fmt, ap);  
    va_end(ap);
    
    if((ret == -1) || (out == NULL)) {
        return context->logstring(context->logrefcon, loglevel, "Memory error, log entry not available");
    }

    ret = context->logstring(context->logrefcon, loglevel, out);
    free(out);
    return ret;
}

